matchtree.go

```diff
@@ -983,14 +991,33 @@ func (d *indexData) newMatchTree(q query.Q, opt matchTreeOpt) (matchTree, error)
 		optCopy := opt
 		optCopy.DisableWordMatchOptimization = true
 
-		subMT, err := d.newMatchTree(s.Expr, optCopy)
+		anchored := false
+		expr := s.Expr
+		switch e := s.Expr.(type) {
+		case *query.Regexp:
+			pattern := e.Regexp.String()
+			if strings.HasPrefix(pattern, "^") && strings.HasSuffix(pattern, "$") {
+				pattern = pattern[1 : len(pattern)-1]
+				parsedPattern, err := syntax.Parse(pattern, e.Regexp.Flags)
+				if err != nil {
+					return nil, err
+				}
+				eCopy := *e
+				eCopy.Regexp = parsedPattern
+				expr = &eCopy
+				anchored = true
+			}
+		}
```

Comment:
When constructing the match tree for symbols, try to strip anchors off the input pattern. We keep track of whether the pattern was anchored, which we then use to post-filter the results so we only return ones that exactly match the symbol boundaries.